routes/courseRoutes.js

```diff
@@ -437,4 +452,245 @@ router.get('/course/delete_all', requireLogin, async (req, res) => {
+		const course = await CourseModel.findById(id);
+
+		// check if courses exist
+		if (!course) {
+			// Handle "course not found" error response here
+			return res.status(404).json({ 'error': errorCodes['E0006'] });
+		}
+
+		res.send(course);
+
+	} catch (error) {
+
+		console.log(error);
+		return res.status(500).json({ 'error': errorCodes['E0003'] });
+	}
+
+})
+
+
+// Get all sections from course
+router.get('/:id/sections', async (req, res) => {
+
+	try {
+		const { id } = req.params;
+
+		// find a course based on it's id
+		const course = await CourseModel.findById(id);
+
+		// check if courses exist
+		if (!course) {
+			// Handle "course not found" error response here
+			return res.status(404).json({ 'error': errorCodes['E0006'] });
+		}
+
+		const sectionsInCourse = course.sections;
+
+		// check if course contains sections
+		if (sectionsInCourse.length === 0) {
+			// Handle "course does not contain sections" error response here
+			return res.status(404).json({ 'error': errorCodes['E0009'] });
+		}
+
+		// check if the sections exists
+		const sectionsList = await SectionModel.find({ '_id': { $in: sectionsInCourse } });
+		if (sectionsList.length === 0) {
+			// Handle "course does not contain sections" error response here
+			return res.status(404).json({ 'error': errorCodes['E0007'] });
+		}
+
+		res.send(sectionsList);
+
+	} catch (error) {
+		console.log(error);
+		return res.status(500).json({ 'error': errorCodes['E0003'] });
+	}
+
+});
+
+// Get a specififc section 
+router.get('/:courseId/sections/:sectionId', async (req, res) => {
+
+	try {
+		const { courseId, sectionId } = req.params;
+
+		const course = await CourseModel.findById(courseId);
+		// check if courses exist
+		if (!course) {
+			// Handle "course not found" error response here
+			return res.status(404).json({ 'error': errorCodes['E0006'] });
+		}
+		// find a specific section within the given course by both IDs
+		const section = await SectionModel.findOne({ parentCourse: courseId, _id: sectionId });
+
+		// check if section exist
+		if (!section) {
+			// Handle "section not found" error response here
+			return res.status(404).json({ 'error': errorCodes['E0008'] });
+		}
+
+		res.send(section);
+
+	} catch (error) {
+		console.log(error);
+		return res.status(500).json({ 'error': errorCodes['E0003'] });
+	}
+
+});
+
+/*
+// Get all excercies from a section *** commented out since we do not use it per 10/10
+router.get("/:courseId/sections/:sectionId/exercises", async (req, res) => {
+
+	try {
+	const { courseId, sectionId } = req.params; 
+
+	// find a specific section within the given course by both IDs
+	const exercises = await ExerciseModel.find({ parentSection: sectionId });
+	res.send(exercises);
+
+	} catch (error) {
+	console.error(error);
+	res.status(500).json({ message: 'Server error' });
+	}
+
+});
+*/
+
+/*** SUBSCRIPTION ROUTES ***/
+
+// Subscribe to course 
+router.post('/:id/subscribe', async (req, res) => {
```

Comment:
Right now our [api guidelines](https://github.com/Educado-App/resources/blob/main/API_GUIDELINES.md#performing-an-action) specify that actions like subscribe should be formatted as e.g. `/couses/some_id:subscribe`.

We should either change the guidelines or follow them. Maybe this is a question for the integration team?

How to add colon in express endpoints:
https://github.com/expressjs/express/issues/3857#issuecomment-589792737

Reply:
it is /courses, check index.js

Reply:
The second option courses/id:subscribe doesn't follow the standard RESTful URL structure, and it might be confusing for other developers who are used to working with RESTful APIs. It also doesn't provide an obvious way to capture the id of the course.

So, in most cases, it's better to stick with the first option (courses/:id/subscribe) for defining routes in a microservices API following RESTful principles.

Reply:
jeg spurgte lige chat GPT: 

" hello, following microservices API guidelines, should I write my route like: router.post('courses/:id/subscribe', async (req, res) => {                                                                                                          
     // or should i do it like this:                                                                                                                                                            router.post('courses/id:subscribe', async (req, res) => {    "

In the context of designing a RESTful API for microservices, the first option you provided is the recommended way to define routes. Here's why:

RESTful URL Structure: REST (Representational State Transfer) is an architectural style that uses a specific URL structure to represent resources and actions. In a RESTful API, resources are typically represented by nouns (like "courses" in your example), and actions are represented by HTTP methods (like POST for creating a subscription). Using the URL path segment courses/:id/subscribe aligns with REST principles, where :id is a placeholder for the specific course's identifier. This makes your API more self-explanatory and easier to understand for developers.

Parameter Parsing: In most web frameworks or libraries (e.g., Express.js for Node.js), you can easily extract the :id parameter from the URL. It allows you to access the id in your route handler, which is crucial for performing actions specific to a course.

Consistency: Following standard conventions, such as RESTful URLs, makes your API consistent and predictable. Developers who work with your API will already be familiar with these conventions and can understand how to interact with it more easily.

The second option courses/id:subscribe doesn't follow the standard RESTful URL structure, and it might be confusing for other developers who are used to working with RESTful APIs. It also doesn't provide an obvious way to capture the id of the course.

So, in most cases, it's better to stick with the first option (courses/:id/subscribe) for defining routes in a microservices API following RESTful principles.

Reply:
Jeg tror ikke subscribe er "action" som beskrivet i API guidelines, siden den opdaterer modellen til user: defor tror jeg det går under denne definition: 
[⛔](https://github.com/Educado-App/resources/blob/main/API_GUIDELINES.md#actions-no-actions-for-crud) DO NOT use an action operation when the operation behavior could reasonably be defined as one of the standard REST Create, Read, Update, Delete, or List operations.

Reply:
altså subscribe tror jeg ikke er action som beskrevet her: [Performing an Action](https://github.com/Educado-App/resources/blob/main/API_GUIDELINES.md#performing-an-action)
The REST specification is used to model the state of a resource, and is primarily intended to handle CRUD (Create, Read, Update, Delete) operations. However, many services require the ability to perform an action on a resource, e.g. getting the thumbnail of an image or rebooting a VM. It is also sometimes useful to perform an action on a collection.

Reply:
og chatGPT mente at courses/:id/subscribe var rigtigt